thoth/adviser/sieves/tensorflow/tf_api.py

```diff
@@ -0,0 +1,131 @@
+    _LINK_API = jl("tf_api")
+    _LINK_NO_API = jl("tf_no_api")
+    _PACKAGES_AFFECTED = ("tensorflow", "tensorflow-gpu", "intel-tensorflow", "tensorflow-cpu")
+
+    _messages_logged = attr.ib(type=Set[str], factory=set, init=False)
+    _no_api_logged = attr.ib(type=bool, default=False, init=False)
+    _acceptable_releases = attr.ib(type=Optional[Set[str]], default=None, init=False)
+
+    def pre_run(self) -> None:
+        """Initialize this pipeline unit before each run."""
+        self._messages_logged.clear()
+        self._acceptable_releases = None
+        self._no_api_logged = False
+        super().pre_run()
+
+    def _pre_compute_releases(self) -> None:
+        """Pre-compute releases that match library usage supplied by the user."""
+        with open(os.path.join(self._DATA_DIR, "tensorflow", "api.json"), "r") as api_file:
+            known_api = json.load(api_file)
+
+        self._acceptable_releases = set()
+        tf_api_used = set(
+            i for i in ((self.context.library_usage.get("report") or {}).get("tensorflow") or [])  # type: ignore
+        )
+        for tf_version, tf_api in known_api.items():
+            if tf_api_used.issubset(tf_api):
+                self._acceptable_releases.add(tf_version)
+
+    @classmethod
+    def should_include(cls, builder_context: "PipelineBuilderContext") -> Optional[Dict[str, Any]]:
+        """Register this pipeline unit for adviser library usage is provided."""
+        if not builder_context.is_adviser_pipeline() or not builder_context.library_usage:
+            return None
+
+        if builder_context.recommendation_type in (RecommendationType.LATEST, RecommendationType.TESTING):
+            # Use any TensorFlow for testing purposes or when resolving latest stack.
+            return None
+
+        if "tensorflow" not in (builder_context.library_usage.get("report") or {}):
+            return None
+
+        # Include this pipeline unit in configurations for various builds.
+        included_units = builder_context.get_included_sieves(cls)
+        if len(included_units) >= len(cls._PACKAGES_AFFECTED):
+            return None
+
+        return {
+            "package_name": cls._PACKAGES_AFFECTED[len(included_units)],
+        }
+
+    def run(self, package_versions: Generator[PackageVersion, None, None]) -> Generator[PackageVersion, None, None]:
+        """Use specific TensorFlow release based on library usage as supplied by the user."""
+        if self._acceptable_releases is None:
+            self._pre_compute_releases()
+
+        if not self._acceptable_releases:
+            if not self._no_api_logged:
+                self._no_api_logged = True
+                msg = "No TensorFlow symbols API found in the database that would match TensorFlow symbols used"
+                self.context.stack_info.append(
+                    {"type": "WARNING", "message": msg, "link": self._LINK_NO_API,}
+                )
+
+            yield from package_versions
```

Comment:
Why do we yield rather than raising an error?

Reply:
Is the assumption that they might be using some un-official part of Tensorflow so erroring would be counter intuitive because then their code wouldn't work as intended?

Reply:
> Is the assumption that they might be using some un-official part of Tensorflow so erroring would be counter intuitive because then their code wouldn't work as intended?

Yes - users can be using some undocumented API. The other use case is a release for which we do not have API symbols - for example <2 releases that do not export symbols in an easily consumable way. In other words, we are not enforcing strict API usage - if we can derive which version users use, we do so. Otherwise we fallback to any version of TF.